src/Microsoft.Azure.WebJobs.Extensions.Kafka/Output/KafkaProducerAsyncCollector.cs

```diff
@@ -78,13 +78,39 @@ private object BuildKafkaDataEvent(T item)
             return new KafkaEventData<T>(item);
         }
 
-        private KafkaEventData<string> BuildKafkaEventData(JObject dataObj)
+        private object BuildKafkaEventData(JObject dataObj)
+        {
+            if (dataObj["Key"] != null && (string)dataObj["Key"] != null)
+            {
+                return BuildKafkaEventDataForKeyValue(dataObj);
+            }
+            else
+            {
+                return BuildKafkaEventDataForValue(dataObj);
+            }
+        }
+
+        private static object BuildKafkaEventDataForValue(JObject dataObj)
         {
             KafkaEventData<string> messageToSend = new KafkaEventData<string>((string)dataObj["Value"]);
             messageToSend.Timestamp = (DateTime)dataObj["Timestamp"];
```

Comment:
If L96-103 and 109-116 is the same, you might do a refactoring of the extract method.

Reply:
There are two reasons for this :-

- Heavy type checking with KafkaEventData class object
- We can't use IKafkaEventData which is the interface as the method of setters are not exposed to public to maintain the mutability for the output binding and immutability for the trigger.

So we can't make changes here thus we need to continue with this